### PR TITLE
Added FLOW support

### LINKED ISSUE #56: Add support for FLOW addresses in encoding library
Here is the encoding library: https://github.com/ensdomains/address-encoder

Example: https://github.com/ensdomains/address-encoder/pull/47/files

Gitter for dev help: https://gitter.im/ethereum/go-ethereum/name-registry

Please provide a reference link so we can check you did the encoding correctly

### PR DESCRIPTION
**Pull request by iRhonin**
Added flow support

## Issue number

Closes #56 

## Description

- Added FLOW encoder/decoder

## List of features added/changed

- FLOW encoder/decoder

## Reference
- [FLOW address type](https://docs.onflow.org/cadence/language/values-and-types/#address-functions)
- [Flow-go sdk](https://github.com/onflow/flow-go/blob/master/model/flow/address.go#L51)

## How Has This Been Tested?
- Using [this](https://docs.onflow.org/cadence/language/values-and-types/#address-functions) example from doc
- Modified first test case to have some zero in start

## Screenshots (if appropriate):
image

## Checklist:

- [x] My code follows the code style of this project.
- [x] My code implements all the required features.

### CODE REVIEW

**Comment by Arachnid** [comment]
I have updated the Travis build to check that entries in the formats, tests, and README lists are in the correct order. Please check your PR and update it to maintain correct ordering in these lists.

**Comment by makoto** [comment]
@iRhonin  Could you resolve conflicts? FYI, I also asked Flow team to review

**Comment by iRhonin** [comment]
@makoto done.

**Comment by cybercent** [comment]
Hi @makoto & @iRhonin 

Are you guys looking to validate a Flow address format or just the length?

While the length verification might match that does not mean the address format is valid.

Also, on Flow addresses have the same visual format for emulator, testnet and mainnet. 
However they are different, meaning one can't use an address that was generated for emulator or testnet on mainnet.

For reference the GoSDK implements address validation here:
https://github.com/onflow/flow-go-sdk/blob/b30af7b11a1b75dcd2549f6640737da1a5a75f79/address.go#L238

**Comment by iRhonin** [comment]
@cybercent Thank you for the response, I think we need to validate the address, not just the length.

File: src/index.ts [hunk 1]
```diff
@@ -629,6 +629,26 @@ function nanoAddressDecoder(data: string): Buffer {
...
+
+// https://github.com/onflow/flow-go/blob/master/model/flow/address.go#L51
+// If b is larger than 8, b will be cropped from the left.
+// If b is smaller than 8, b will be appended by zeroes at the front.
+function flowEncode(data: Buffer): string {
+  const AddressLength = 8;
+  let addrBytes = Buffer.alloc(AddressLength, 0x00);
+
+  if (data.length > 8) {
+    addrBytes = data.slice(-8);
```
Comment (turbolent) [inline src/index.ts:645]: Maybe replace the magic numbers with the constant defined above:

```suggestion
  if (data.length > AddressLength) {
    addrBytes = data.slice(-AddressLength);
```

**Comment by makoto** [comment]
```
Also, on Flow addresses have the same visual format for emulator, testnet and mainnet.
However they are different, meaning one can't use an address that was generated for emulator or testnet on mainnet.
```

Can you encode and decode back the address? If yes, we should check whether addresses are mainnet, testnet, or emulator and reject anything else. Some chain cannot derive chain info from the encoded format (I think Polkadot and its testnet Kusama has SLIP44 id for each net for that reason). If FLOW is similar to that, then we should only check and allow mainnet.

**Comment by psiemens** [comment]
> Some chain cannot derive chain info from the encoded format 

You can't directly derive the chain ID from the address, but you can check for validity against each of the 3 networks (emulator, testnet, mainnet).

@turbolent I can't recall if we've ported the linear code logic to JavaScript yet. Do you know if that exists yet?

**Comment by turbolent** [comment]
@psiemens I don't think so. I ported it here to TypeScript: https://gist.github.com/turbolent/906d651c8776bf63e3e8c4846bb1c363

@orodio Can we port this to JS and include it in the JS SDK?

**Comment by makoto** [comment]
@turbolent we are very sensitive about bloating this library so we tend to port it to this library itself. Since this is already typescript, I assume it is going to be a simple copy&paste?

**Comment by makoto** [comment]
Hello, @iRhonin is this ready or still wip?

**Comment by iRhonin** [comment]
> Hello, @iRhonin is this ready or still wip?

Hi @makoto, no I waiting for more information from the flow team, the gist is private.

**Comment by psiemens** [comment]
@iRhonin, oops didn't realize that was a private gist!

Here's the contents, feel free to copy/paste if that's better for this library.

```ts
type Address = bigint

// [n,k,d]-Linear code parameters
// The linear code used in the account addressing is a [64,45,7]
// It generates a [64,45]-code, which is the space of Flow account addresses.
//
// n is the size of the code words in bits,
// which is also the size of the account addresses in bits.
const linearCodeN = 64

// Columns of the parity-check matrix H of the [64,45]-code used for Flow addresses.
// H is a (n x p) matrix with coefficients in GF(2), each column is converted into
// a big endian integer representation of the GF(2) column vector.
// H is used to verify a code word is a valid account address.
const parityCheckMatrixColumns: bigint[] = [
	0x00001n, 0x00002n, 0x00004n, 0x00008n,
	0x00010n, 0x00020n, 0x00040n, 0x00080n,
	0x00100n, 0x00200n, 0x00400n, 0x00800n,
	0x01000n, 0x02000n, 0x04000n, 0x08000n,
	0x10000n, 0x20000n, 0x40000n, 0x7328dn,
	0x6689an, 0x6112fn, 0x6084bn, 0x433fdn,
	0x42aabn, 0x41951n, 0x233cen, 0x22a81n,
	0x21948n, 0x1ef60n, 0x1decan, 0x1c639n,
	0x1bdd8n, 0x1a535n, 0x194acn, 0x18c46n,
	0x1632bn, 0x1529bn, 0x14a43n, 0x13184n,
	0x12942n, 0x118c1n, 0x0f812n, 0x0e027n,
	0x0d00en, 0x0c83cn, 0x0b01dn, 0x0a831n,
	0x0982bn, 0x07034n, 0x0682an, 0x05819n,
	0x03807n, 0x007d2n, 0x00727n, 0x0068en,
	0x0067cn, 0x0059dn, 0x004ebn, 0x003b4n,
	0x0036an, 0x002d9n, 0x001c7n, 0x0003fn,
]


function isValidAddress(address: Address, chain: ChainID): boolean {
	let codeWord: bigint = address
	codeWord ^= chain.codeword

	if (codeWord === 0n) {
		return false
	}

	// Multiply the code word GF(2)-vector by the parity-check matrix
	let parity = 0n
	for (let i = 0; i < linearCodeN; i++) {
		if ((codeWord & 1n) === 1n) {
			parity ^= parityCheckMatrixColumns[i]
		}
		codeWord >>= 1n
	}
	return (parity === 0n) && (codeWord === 0n)
}

class ChainID {

	constructor(readonly codeword: bigint) {}

	static mainnet = new ChainID(0n)
	static testnet = new ChainID(0x6834ba37b3980209n)
	static emulator = new ChainID(0x1cb159857af02018n)
}

const invalidCodeWord = 0xab2ae42382900010n

console.log(isValidAddress(invalidCodeWord, ChainID.mainnet))
console.log(isValidAddress(0x7e60df042a9c0868n, ChainID.testnet))
```

**Comment by iRhonin** [comment]
Hi @psiemens, thank you for the code, almost copy-paste it.
Can you please send me a valid flow address in mainnet?

**Comment by makoto** [comment]
@iRhonin there is a valid address at https://docs.onflow.org/cadence/language/values-and-types/#addresses

**Comment by turbolent** [comment]
Secret gists are still public, they're just not listed on the profile page.

As for valid addresses, the link above is not useful, as it only documents addresses in Cadence itself, which has no knowledge of Flow networks.

Instead, for example consider the address of the fungible token contract: https://docs.onflow.org/protocol/core-contracts/fungible-token/

**Comment by makoto** [comment]
Hello, how is it progressing? Need any more info/help?

**Comment by iRhonin** [comment]
> Hello, how is it progressing? Need any more info/help?

I am going to fix this at the weekend.

**Comment by iRhonin** [comment]
Hi @makoto,
I copy/paste the code and replaced the test case with a real one.
But I had to bump nodejs version to 14 and es to 2020 to use BigInt notation. Is this okay?

**Comment by makoto** [comment]
`But I had to bump nodejs version to 14 and es to 2020 to use BigInt notation. Is this okay?` I need to test. `ens-app` runs on node14, but `ui` is still on node 10 and last time I upgraded it failed so may cause some issue.

**Comment by makoto** [comment]
BTW, tests are failing

**Comment by iRhonin** [comment]
@makoto I have downgraded to node 10, please checkout the code.

**Pull request merged**